services/tasks/src/main/java/org/hyperledger/besu/services/tasks/InMemoryTasksPriorityQueues.java

```diff
@@ -0,0 +1,169 @@
+ * an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the
+ * specific language governing permissions and limitations under the License.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+package org.hyperledger.besu.services.tasks;
+
+import java.util.ArrayList;
+import java.util.Comparator;
+import java.util.HashSet;
+import java.util.List;
+import java.util.PriorityQueue;
+import java.util.Queue;
+import java.util.Set;
+import java.util.concurrent.atomic.AtomicBoolean;
+
+import org.jetbrains.annotations.NotNull;
+
+public class InMemoryTasksPriorityQueues<T extends TasksPriorityProvider>
+    implements TaskCollection<T> {
+  private final List<PriorityQueue<T>> internalQueues = new ArrayList<>(17);
+  private final Set<InMemoryTask<T>> unfinishedOutstandingTasks = new HashSet<>();
+  private final AtomicBoolean closed = new AtomicBoolean(false);
+
+  public InMemoryTasksPriorityQueues() {
+    clearInternalQueues();
+  }
+
+  private void clearInternalQueues() {
+    internalQueues.clear();
+    for (int i = 0; i < 16; i++) {
+      internalQueues.add(newEmptyQueue());
+    }
+  }
+
+  @NotNull
+  private PriorityQueue<T> newEmptyQueue() {
+    return new PriorityQueue<>(Comparator.comparingLong(TasksPriorityProvider::getPriority));
+  }
+
+  @Override
+  public synchronized void add(final T taskData) {
+    assertNotClosed();
+    var dequeue = findQueue(taskData.getDepth());
+    dequeue.add(taskData);
+  }
+
+  private PriorityQueue<T> findQueue(final int priority) {
+    while (priority + 1 > internalQueues.size()) {
+      internalQueues.add(newEmptyQueue());
+    }
+    return internalQueues.get(priority);
+  }
+
+  @Override
+  public synchronized Task<T> remove() {
+    assertNotClosed();
+    final Queue<T> lastNonEmptyQueue = findLastNonEmptyQueue();
+    if (lastNonEmptyQueue.isEmpty()) {
+      return null;
+    }
+    T data = lastNonEmptyQueue.remove();
+    InMemoryTask<T> task = new InMemoryTask<>(this, data);
+    unfinishedOutstandingTasks.add(task);
+    return task;
+  }
+
+  private Queue<T> findLastNonEmptyQueue() {
+    for (int i = internalQueues.size() - 1; i > 0; i--) {
+      final Queue<T> queue = internalQueues.get(i);
+      if (!queue.isEmpty()) {
+        return queue;
+      }
+    }
+    return internalQueues.get(0);
+  }
+
+  @Override
+  public synchronized long size() {
+    return internalQueues.stream().mapToInt(Queue::size).sum();
+  }
+
+  @Override
+  public synchronized boolean isEmpty() {
+    return findLastNonEmptyQueue().isEmpty();
+  }
+
+  @Override
+  public synchronized void clear() {
+    assertNotClosed();
+
+    unfinishedOutstandingTasks.clear();
+    clearInternalQueues();
+  }
+
+  @Override
+  public synchronized boolean allTasksCompleted() {
+    return isEmpty() && unfinishedOutstandingTasks.isEmpty();
+  }
+
+  @Override
+  public synchronized void close() {
+    closed.set(true);
```

Comment:
Why is close not checking the `closed` bit?

Reply:
also, why is close not handling the `unfinishedOutstandingTasks` set? Do you want to wait for them to complete to return? What's the handling of this?

Reply:
I will look into this. I copied over what was used in other implementations, might be worth double-checking what to do correctly here.

Reply:
You are right, unfinishedOutstandingTasks should be cleared when close is called.  I am changing it. Also, it makes sense to only clear the collections once.